### ISSUE TITLE
10x speed up for `transform.PiecewiseAffineTransform`

### ISSUE THREAD

**Issue by svepe** [scikit-image/scikit-image#6864]
### Description:

I've run into some issues with `transform.PiecewiseAffineTransform` being quite slow and came up with a 10x speed up. The idea is to replace the slow [iteration over simplices](https://github.com/scikit-image/scikit-image/blob/ba481a69fc0edabb1dcf5c17c49adfc8657d00c7/skimage/transform/_geometric.py#L1042) with a numpy operation as implemented in the `FastPiecewiseAffineTransform` bellow.

```python
import time
import numpy as np
from skimage.transform import PiecewiseAffineTransform, warp


class FastPiecewiseAffineTransform(PiecewiseAffineTransform):
    def __call__(self, coords):
        coords = np.asarray(coords)

        simplex = self._tesselation.find_simplex(coords)

        affines = np.array(
            [self.affines[i].params for i in range(len(self._tesselation.simplices))]
        )[simplex]

        pts = np.c_[coords, np.ones((coords.shape[0], 1))]

        result = np.einsum("ij,ikj->ik", pts, affines)
        result[simplex == -1, :] = -1

        return result


def main():
    # Create dummy data for testing
    src = np.random.rand(500, 2)
    dst = src + 0.1 * np.random.rand(500, 2)

    size = 2000
    src *= size
    dst *= size
    image = np.random.rand(size, size, 3)

    # Run the standard PiecewiseAffine transform
    tf = PiecewiseAffineTransform()
    tf.estimate(src, dst)
    start = time.time()
    slow = warp(image, tf)
    print(f"    PiecewiseAffineTransform took: {time.time() - start} s")

    # Run the FastPiecewiseAffine transform
    tf = FastPiecewiseAffineTransform()
    tf.estimate(src, dst)
    start = time.time()
    fast = warp(image, tf)
    print(f"FastPiecewiseAffineTransform took: {time.time() - start} s")

    # Ensure images are identical
    assert np.allclose(fast, slow)


if __name__ == "__main__":
    main()
```
Here is the output:

```shell
    PiecewiseAffineTransform took: 16.557838439941406 s
FastPiecewiseAffineTransform took: 1.2521417140960693 s
```

This comes at the cost of not following the standard transform API as the underlying matrices of the individual affine transforms are accessed and I am not sure what the implications of that are. In any case, I'm happy to submit a PR if this change makes sense from an architectural perspective.

**Comment by stefanv** [comment]
> This comes at the cost of not following the standard transform API as the underlying matrices of the individual affine transforms are accessed

Could you expand on this a bit? It wasn't immediately apparent to me.

**Comment by svepe** [comment]
The [code](https://github.com/scikit-image/scikit-image/blob/ba481a69fc0edabb1dcf5c17c49adfc8657d00c7/skimage/transform/_geometric.py#L1048) currently applies each of the `AffineTransform`s by using their `__call__` method which is the standard way of doing it. The proposed faster version uses the underlying `params` matrix i.e.
```python
        affines = np.array(
            [self.affines[i].params for i in range(len(self._tesselation.simplices))]
        )[simplex]
``` 
and then applies the transformations by carrying out the matrix multiplication with `einsum`.

**Comment by 23pointsNorth** [comment]
Any updates on this? It's being heavily used in all of ML e.g. [Albumentations](https://github.com/albumentations-team/albumentations/blob/master/albumentations/augmentations/geometric/transforms.py#L940) and would make things much faster and more energy efficient.

**Comment by lagru** [comment]
I don't see an issue with accessing `self.affines[i].params` in `__call__`. The attribute is part of the classes public API of `AffineTransform`. I agree that it doesn't seem as elegant, but honestly the whole transform API is something we have discussed refactoring in some bright future.

To be careful I would probably check if the instances of `self.affines` is what we expect them to be. Otherwise we might raise a warning or fall back to the old behavior. But that can be discussed out later. So in the meantime, please go ahead and make a PR. Thanks for the great snippet by the way! :pray:

**Comment by stefanv** [comment]
Agreed, using internal API during calculation is perfectly acceptable.

**Comment by lagru** [comment]
> In any case, I'm happy to submit a PR if this change makes sense from an architectural perspective.

@svepe if it's still convenient for you, please go ahead and feel welcome to submit a PR! :D

**Comment by kippapollo** [comment]
Thanks for your implement to speed up PiecewiseAffineTransform.
I am sorry but could you know how can I run PiecewiseAffineTransform on CUDA?

**Comment by svepe** [comment]
> @svepe Thanks for your implement to speed up PiecewiseAffineTransform. I am sorry but could you know how can I run PiecewiseAffineTransform on CUDA?

@kippapollo If you use something like CuPy then you can add your implementation, simillar to the `FastPiecewiseAffineTransform` class above using cupy calls instead of numpy ones, but I am not sure how much faster it will be given that the matrices are small and so the overhead of moving them to the GPU will probably result in an overall longer computation.

**Comment by kippapollo** [comment]
@svepe Thanks for you reply.
What is mean the matrices are small and why is it the overhead of moving them to the GPU?
